File: http_binding.go
package plugin_sdk

import (
	"encoding/json"
	"fmt"
	"strings"
	"unicode"
	"unicode/utf8"

	pb "github.com/torana-edge/torana-plugin-sdk/pb/v1"
	"github.com/torana-edge/torana-plugin-sdk/strictjson"
)

// HTTPConversationBinding is host-owned MCP evidence, not caller input.
// ConversationID identifies the session; it does not identify a side thread
// or reproduce a plugin's own state-key derivation.
type HTTPConversationBinding struct {
	Bound          bool
	ConversationID string
	CallID         string
}

// HTTPConversation reads the binding injected by Torana into an HTTP hook.
// present=false means ordinary HTTP, with no MCP observation. present=true
// and Bound=false means explicitly unbound; neither permits scoped state access.
// This validates the envelope, not its provenance: only use the request supplied
// to OnHTTPRequest. Never apply it to model arguments or caller-created headers.
func HTTPConversation(req *pb.HttpRequest) (HTTPConversationBinding, bool, error) {
	var empty HTTPConversationBinding
	if req == nil || len(req.HeadersJson) == 0 {
		return empty, false, nil
	}
	object, err := strictjson.DecodeObject(req.HeadersJson)
	if err != nil || object == nil {
		return empty, false, fmt.Errorf("torana: invalid HTTP header envelope")
	}
	values := map[string]string{}
	for name, raw := range object {
		key := strings.ToLower(name)
		if key != "x-torana-mcp-binding" && key != "x-torana-conversation-id" && key != "x-torana-tool-use-id" {
			continue
		}
		if _, exists := values[key]; exists {
			return empty, false, fmt.Errorf("torana: duplicate HTTP binding header")
		}
		encoded, err := json.Marshal(raw)
		var list []string
		if err != nil || json.Unmarshal(encoded, &list) != nil || len(list) != 1 {
			return empty, false, fmt.Errorf("torana: HTTP binding header needs one value")
		}
		values[key] = list[0]
	}
	status, present := values["x-torana-mcp-binding"]
	conversation, hasConversation := values["x-torana-conversation-id"]
	call, hasCall := values["x-torana-tool-use-id"]
	if !present || status == "unbound" {
		if hasConversation || hasCall {
			return empty, false, fmt.Errorf("torana: unbound HTTP call cannot carry identities")
		}
		return empty, present, nil
	}
	valid := func(value string) bool {
		return value != "" && len(value) <= 256 && utf8.ValidString(value) && strings.IndexFunc(value, unicode.IsControl) < 0
	}
	if status != "bound" || !valid(conversation) || !valid(call) {
		return empty, false, fmt.Errorf("torana: invalid HTTP conversation binding")
	}
	return HTTPConversationBinding{Bound: true, ConversationID: conversation, CallID: call}, true, nil
}

File: rust/torana-plugin-sdk/src/http_binding.rs
use crate::{pbv1, strict_json};

/// Host-owned MCP evidence. A session identity is not a plugin's thread/state key.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct HttpConversationBinding {
    pub bound: bool,
    pub conversation_id: String,
    pub call_id: String,
}

/// Read only from the request supplied to Plugin::on_http. This validates the
/// envelope, not provenance; model arguments and caller-created headers are not
/// trusted evidence. None means ordinary HTTP; Some with bound=false means MCP
/// explicitly reported an unbound call. Neither authorizes scoped state access.
pub fn http_conversation(
    req: &pbv1::HttpRequest,
) -> Result<Option<HttpConversationBinding>, String> {
    if req.headers_json.is_empty() {
        return Ok(None);
    }
    let raw = strict_json(&req.headers_json).map_err(|_| "invalid HTTP header envelope")?;
    let object = raw.as_object().ok_or("invalid HTTP header envelope")?;
    let mut values = std::collections::BTreeMap::new();
    for (name, value) in object {
        let key = name.to_ascii_lowercase();
        if !matches!(
            key.as_str(),
            "x-torana-mcp-binding" | "x-torana-conversation-id" | "x-torana-tool-use-id"
        ) {
            continue;
        }
        let list = value
            .as_array()
            .ok_or("HTTP binding header needs one value")?;
        if list.len() != 1 {
            return Err("HTTP binding header needs one value".into());
        }
        let value = list[0]
            .as_str()
            .ok_or("HTTP binding header needs one value")?;
        if values.insert(key, value).is_some() {
            return Err("duplicate HTTP binding header".into());
        }
    }
    let status = values.get("x-torana-mcp-binding").copied();
    let conversation = values.get("x-torana-conversation-id").copied();
    let call = values.get("x-torana-tool-use-id").copied();
    if status.is_none() || status == Some("unbound") {
        if conversation.is_some() || call.is_some() {
            return Err("unbound HTTP call cannot carry identities".into());
        }
        return Ok(status.map(|_| HttpConversationBinding {
            bound: false,
            conversation_id: String::new(),
            call_id: String::new(),
        }));
    }
    let valid = |value: Option<&str>| {
        value.is_some_and(|v| !v.is_empty() && v.len() <= 256 && !v.chars().any(char::is_control))
    };
    if status != Some("bound") || !valid(conversation) || !valid(call) {
        return Err("invalid HTTP conversation binding".into());
    }
    Ok(Some(HttpConversationBinding {
        bound: true,
        conversation_id: conversation.unwrap().into(),
        call_id: call.unwrap().into(),
    }))
}

#[cfg(test)]
mod tests {
    use super::*;
    fn read(raw: &str) -> Result<Option<HttpConversationBinding>, String> {
        http_conversation(&pbv1::HttpRequest {
            headers_json: raw.as_bytes().into(),
            ..Default::default()
        })
    }
    #[test]
    fn absent_unbound_and_bound_are_distinct() {
        assert_eq!(read("").unwrap(), None);
        assert_eq!(read("{}").unwrap(), None);
        assert!(
            !read(r#"{"X-Torana-MCP-Binding":["unbound"]}"#)
                .unwrap()
                .unwrap()
                .bound
        );
        let binding = read(r#"{"x-torana-mcp-binding":["bound"],"X-Torana-Conversation-Id":["session"],"X-Torana-Tool-Use-Id":["call"]}"#).unwrap().unwrap();
        assert_eq!(
            binding,
            HttpConversationBinding {
                bound: true,
                conversation_id: "session".into(),
                call_id: "call".into()
            }
        );
    }
    #[test]
    fn refuses_ambiguous_and_invalid_evidence() {
        for raw in [
            "null",
            "[]",
            "{} {}",
            r#"{"X-Torana-MCP-Binding":["bound"],"X-Torana-MCP-Binding":["unbound"]}"#,
            r#"{"X-Torana-MCP-Binding":["unbound"],"x-torana-mcp-binding":["unbound"]}"#,
            r#"{"X-Torana-MCP-Binding":[]}"#,
            r#"{"X-Torana-MCP-Binding":["bound","unbound"]}"#,
            r#"{"X-Torana-MCP-Binding":["unknown"]}"#,
            r#"{"X-Torana-MCP-Binding":["bound"]}"#,
            r#"{"X-Torana-Conversation-Id":["private-session"]}"#,
            r#"{"X-Torana-MCP-Binding":["unbound"],"X-Torana-Tool-Use-Id":["private-call"]}"#,
            r#"{"X-Torana-MCP-Binding":["bound"],"X-Torana-Conversation-Id":["session\n"],"X-Torana-Tool-Use-Id":["call"]}"#,
        ] {
            assert!(read(raw).is_err());
        }
        assert!(read(&format!(r#"{{"X-Torana-MCP-Binding":["bound"],"X-Torana-Conversation-Id":["{}"],"X-Torana-Tool-Use-Id":["call"]}}"#, "a".repeat(257))).is_err());
    }
}
